### ISSUE TITLE
Kubeapps error in syncing with karpenter(public ecr) due to version mismatch

### ISSUE THREAD

**Issue by bala151187** [vmware-tanzu/kubeapps#6160]
Kubeapps is unable to sync with public ecr(AWS) for karpenter because of the version mismatch with checksum

kube apps sync log 
`I0331 18:15:40.393944       1 sync.go:77] Last checksum: 
E0331 18:15:49.031187       1 utils.go:585] failed to pull chart. Got public.ecr.aws/karpenter/karpenter-crd:0.27.1: not found
E0331 18:15:49.031259       1 utils.go:585] failed to pull chart. Got public.ecr.aws/karpenter/karpenter:0.27.1: not found
I0331 18:15:49.031291       1 sync.go:99] No charts in repository to be synced, repo.URL= https://public.ecr.aws/karpenter
time="2023-03-31T18:15:49Z" level=info msg="trying next host - response was http.StatusNotFound" host=public.ecr.aws
time="2023-03-31T18:15:49Z" level=info msg="trying next host - response was http.StatusNotFound" host=public.ecr.aws`

AWS ECR tags for karpenter chart

![Screen Shot 2023-03-31 at 2 38 44 PM](https://user-images.githubusercontent.com/5027627/229203138-9ae5bed4-91a8-436e-9a96-c95d83e0914b.png)

ECR repo got the tag like `v0.27.1` but kubeapps checksum says `0.27.1` that why 404 error 

If Kubeapps can provide way to specify the version this can be handled i think

**Comment by absoludity** [comment]
Hi @bala151187 . Woah - that is a confusing situation. So the Helm docs say explicitly that the [Helm chart version is a semantic version](https://helm.sh/docs/topics/charts/#charts-and-versioning) (which [does not allow a v in the major version](https://semver.org/#is-v123-a-semantic-version) - the `v` is used for lots of tags in git but is a prefix to the semver, not part of it).

And in fact, checking the Chart.yaml of the karpenter Helm chart, is indeed follows this:

https://github.com/aws/karpenter/blob/31799f5c73cd92282030c3f4250d5513fb6fd636/charts/karpenter/Chart.yaml#L5

Yet the images are tagged not with the (semver) version, but with the version tag (including the `v` prefix), and because the Helm client just gets you to specify the chart image, this works regardless.

This does not seem to be intended in Helm ([see the example here](https://helm.sh/blog/storing-charts-in-oci/#working-with-charts-in-registries)), nor in the [Amazon ECR docs for pushing Helm charts to ECR](https://docs.aws.amazon.com/AmazonECR/latest/userguide/push-oci-artifact.html). Everything I see seems to indicate that the expectation is still, even with OCI charts.that the chart name includes the semver version (without the prefix).

Even the instructions for pushing the helm chart to an OCI repository seem to imply that the Helm tool ensures this is the case:

```
$ helm push demo-0.1.0.tgz oci://r.example.com/myuser
Pushed: r.example.com/myuser/demo:0.1.0
```
(note, the `tgz` file is created by the `helm package` command, based on the version in the Chart.yaml, which is semver even for Karpenter).

So I'm not sure how the project is even creating this situation? Do you have more info?

**Comment by bala151187** [comment]
@absoludity .. Thanks for the reply .. Karpenter doc says 
`helm upgrade --install --namespace karpenter --create-namespace \
  karpenter oci://public.ecr.aws/karpenter/karpenter \
  **--version v0.27.1** \
  --set serviceAccount.annotations.eks\.amazonaws\.com/role-arn=${KARPENTER_IAM_ROLE_ARN} \
  --set settings.aws.clusterName=${CLUSTER_NAME} \
  --set settings.aws.clusterEndpoint=${CLUSTER_ENDPOINT} \
  --set settings.aws.defaultInstanceProfile=KarpenterNodeInstanceProfile-${CLUSTER_NAME} \
  --set settings.aws.interruptionQueueName=${CLUSTER_NAME} \
  --wait`
[link](https://github.com/aws/karpenter/tree/31799f5c73cd92282030c3f4250d5513fb6fd636/charts/karpenter)

Eventhough the karpenter [chart.yml](https://github.com/aws/karpenter/blob/31799f5c73cd92282030c3f4250d5513fb6fd636/charts/karpenter/Chart.yaml) Doesn't have `**v**` prefixed .. Karpenter expecting the chart version like this **v0.27.1** . That's why i'm confused too .. 

This is how i have actually configured . 
![Screen Shot 2023-04-03 at 10 10 52 AM](https://user-images.githubusercontent.com/5027627/229535147-dfeac469-b840-491f-9001-4a18abd9581c.png)

**Comment by absoludity** [comment]
> @absoludity .. Thanks for the reply .. Karpenter doc says [snip]

Yeah, and that works for helm OCI charts (only) because the `v0.27.1` is actually the image tag in this case. I just don't understand why they are tagging their images with the non-semantic version different to the Chart.yaml (or how they are doing so).

> 
> Eventhough the karpenter [chart.yml](https://github.com/aws/karpenter/blob/31799f5c73cd92282030c3f4250d5513fb6fd636/charts/karpenter/Chart.yaml) Doesn't have `**v**` prefixed .. Karpenter expecting the chart version like this **v0.27.1** . That's why i'm confused too ..

Yeah, I've added a bit more specific info to the issue you created there. Let's see what they say. I'm generally not keen for us to support non-standard decisions like this if no one else is doing so. On the other hand, if it's something lots of people who are using Helm's OCI support are doing, we can look at supporting it. I'll mark this as awaiting more evidence until we hear back.

**Comment by absoludity** [comment]
This is still waiting for further input on the upstream karpenter issue aws/karpenter#3689 , though looks unlikely it'll be prioritised there. The fundamental problem is that the Chart.yaml *correctly* specifies the version as `0.27.1` but the chart image is published not with `0.21.1` but with `v0.21.1` so Kubeapps is unable to sync the image.

We (kubeapps) could respond to an initial 404 when syncing an OCI chart by checking a tag prefixed with `v` (when the tag is not already prefixed), as a work-around.
